[FTR] use sencha’s native ant injection points for building

bwbohl opened the pull request:
## Description, Context and related Issue

This pull request incorporates all commands executed by `build.sh` into `build.xml`:

* introducing a new target for dependency check to mimic the dependency check in build.sh
* separate resource copying and zipping for XAR
* use sencha ant injection points for managing the build order
* use -clean injection point to avoid deletion of `build` directory but clean its contents instead (this is good when bind mounting to a Docker image or using VSCode exist extension for syncing build to the database)
* keep the "xar" target to keep build.sh functional

With this PR `build.sh` becomes obsolete becasue `sencha app build [production | testing | …]` will create the exact same result including the XAR creation.

Building across different platforms might profit from that, `build.bat`should become obsolete, too.


<!--- This project only accepts pull requests related to open issues. Please link to the issue here: -->
Refs https://github.com/Edirom/Edirom-Online-Frontend/issues/{ISSUE_NUMBER}

## How Has This Been Tested?
built using both build.sh and sencha app build and diffing directories

## Types of changes
<!--- What types of changes does your code introduce? Please DELETE options that are not relevant. -->
- Improvement
- Refactoring

## Overview
<!--- Go over all the following points, and DELETE options that are not relevant. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- I have performed a self-review of my code, according to the [style guide](https://github.com/Edirom/Edirom-Online/blob/develop/STYLE-GUIDE.md)
- I have read the [CONTRIBUTING](https://github.com/Edirom/Edirom-Online-Frontend/blob/develop/CONTRIBUTING.md) document.

bwbohl committed:
build.xml: remove trailing whitespaces

bwbohl committed:
build.xml: move copying xar resources to new task

bwbohl committed:
build.xml: create target for Linux bash dependency check (as in build.sh)

bwbohl committed:
build.xml: make dependency-check fail on error

bwbohl committed:
build.xml: add dependencies to xar target

With this commit, `ant xar` will create the same results as previously running` build.sh`.

bwbohl committed:
build.xml: add sencha cmd -afer-page injection target

With this the build result of `sencha app build` will be exactly the same as running `build.sh` (except for creating the xar archive)

bwbohl committed:
build.xml: override sencha -clean target to delete build directory contents only

bwbohl committed:
build.xml: use -before-build injection point to clean

bwbohl committed:
build.xml: make `build.sh` and `sencha app build` work alongside eachother

bwbohl commented on build.xml:
```diff
@@ -24,6 +24,30 @@
+    
+    <target name="-after-init">
+        <antcall target="check-dependencies" />
+    </target>
```
uses the sencha ant -after-init extension point to call our dependency check

bwbohl commented on build.xml:
```diff
@@ -68,13 +92,28 @@
+    
+    <target name="xar" depends="check-dependencies, clean, build, download-euryanthe, inject-properties, copy-xar-resources">
+        <antcall target="-xar" />
+    </target>
```
Our xar target was split into copy-xar-resources (above) and redefined here to execute all the dependencies before calling -xar target which now only does the zipping

daniel-jettka commented:
> I still kept the build.sh the way it was for comparison and compatibility reasons as described in the original description.
> 
> To test please
> 
>     1. run the build as you are used to with build.sh -d

* removed build dir, but had to create it again because it's not auto-created in build process
* `sh build.sh -d`
* done.

>     2. then rename the build directory to preserve it for comparison

* `mv build build-1`
* done.
 
>     3. now launch bwbohl/sencha-cmd with an interactive shell as described in its repo

* `docker run --rm -it -v $(pwd):/app --name sencha ghcr.io/bwbohl/sencha-cmd:latest`
* done.
 
>     4. build endirom using sencha app build with the same parameter as before (e.g. production, testing, …)

* `root@2ef5680d6ae8:/app# sencha app build` -> build dir missing
* `mkdir build`
* `root@2ef5680d6ae8:/app# sencha app build`
* done.

>     5. back in your OS diff the build directories from steps 2 and 4

* done.

> Results should be the same

* YES, they are exactly the same.

daniel-jettka reviewed (changes requested):
Can you add creation of build dir if it does not exist?

I would approve then

bwbohl committed:
build.xml: prevent clean task failure on non-existing build direcotry

The ant `FileSet` task fails on a missing directory by default; setting `@erroronmissingdir`to `false` prevents that behaviour.

bwbohl committed:
build.xml: add description for check-dependencies target

bwbohl committed:
build.xml: add general hint to Sencha Cmd documentation

bwbohl committed:
build.xml: apply consistent formatting

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl committed:
remove build.sh and build.bat

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

daniel-jettka reviewed (dismissed):
I think it looks good, thanks!

Only thing is that README needs an update now in this section -> https://github.com/baz-ga/Edirom-Online-Frontend/tree/ftr/native-ant-build?tab=readme-ov-file#building-locally

fmacca reviewed (changes requested):
It works well on windows with the ediBuild docker image.

I think this line in the Dockerfile should also be changed to build the whole application with docker compose: https://github.com/baz-ga/Edirom-Online-Frontend/blob/187ae5803d96ffdaee524a5bc763f8d6e11ff4c7/Dockerfile#L69 --> `sencha app build`

bwbohl commented:
Thanks for your reviews and the good catches!

I should also change the default target of the `build.xml` from xar to `build`, I think.

The new build command would actually be two possibilities:

1. `sencha app build` (with optional parameters, e.g., `testing`, `production`, etc., or
2. their sencha ant-impl.xml equivalents, e.g.: `ant build` or `ant build-testing`, etc.

Should I put both options into the documentation, or is there one we prefer?

bwbohl committed:
build.xml: remove deprecated target `xar`

bwbohl committed:
README: remove trailing whitespaces

bwbohl committed:
README: update build instructions and add further reading references

bwbohl committed:
README: fix broken sentence and link

bwbohl committed:
Dockerfile: switch to sencha app build command

bwbohl committed:
Dockerfile: allow submitting Sench build environment, e.g., production / testing

github-actions[bot] committed:
Update CITATION.cff commit and date-released [skip ci]

bwbohl committed:
Merge branch 'develop' into ftr/native-ant-build

fmacca reviewed (approved):
The dockerfile should then be changed also in the main repo here: https://github.com/Edirom/Edirom-Online/blob/fcddbe33f6121d57304245f4abe73e319671e5b9/frontend/Dockerfile#L73
All the rest looks good to me now!

peterstadler reviewed (approved):
I tested this with the Dockerfile as well as with the `ghcr.io/bwbohl/sencha-cmd:2.1.0` image. Both worked nicely, many thanks!
